Feature: Add Subgraph step-by-step guide (#25)
aragon/osx-plugin-template-hardhat: USAGE_GUIDE.md

   If you deploy upcoming versions of your plugin, you must increment the build or release number accordingly (see [our docs on versioning your plugin](https://devs.aragon.org/docs/osx/how-to-guides/plugin-development/publication/versioning)).

> ### Contract Testing

Reviewer: Please create a `Contract` heading as it was done before and do not repeat the term `Contract` in every sub-heading.

Reviewer 2: addressed in [6e1e9e1](https://github.com/aragon/osx-plugin-template-hardhat/pull/25/commits/6e1e9e1354f20ae7c695e0ea31dd55841e73ca5f)

Editing this variables you can change how the subgraph is built and deployed.

## Creating a Subgraph

### `manifest/subgraph.placeholder.yaml`

The first step to create a subgraph is to create or edit the `subgraph.placeholder.yaml` file. This file is located in the `manifest` folder. This file contains the template for the `subgraph.yaml` file. This file is used by the `build-manifest.sh` script to generate the `subgraph.yaml` file. You can find more information about the `subgraph.yaml` file [here](https://thegraph.com/docs/define-a-subgraph#the-subgraph-manifest).

You will see that the `subgraph.placeholder.yaml` file contains some placeholders that will be replaced by the `build-manifest.sh` script. These placeholders are the following:

- `{{dataSources.PluginSetupProcessors}}`: The object containing the processors for the plugin setup events.
- `{{startBlock}}`: The block number where the subgraph will start indexing.
- `{{address}}`: The address of the contract that will be indexed.
- `{{network}}`: The network where the contract is deployed.

> This placeholder are substituted using moustache. You can find more information about it [here](https://mustache.github.io/).

Reviewer: ```suggestion
These placeholders are substituted using [mustache](https://mustache.github.io/).
```

Here you can add your custom entities and how they are related to each other. You can find more information about the schema [here](https://thegraph.com/docs/define-a-subgraph#defining-the-schema).

## Handlers

Events handler functions are associated with the events in the `subgraph.yaml` file and defined in the `src` folder.

The `src` folder contains a `plugin` and `osx` subfolder.
In `plugin.ts` inside the `plugin` folder, you can add plugin-related event handlers.
In `pluginSetupProcessor.ts` inside the `osx` folder, you can add plugin-setup-processor-related event handlers that your plugin might require.

For the `MyPlugin` example, two handlers are already provided:

- `handleInstallationPrepared`: Handles the `InstallationPrepared` event. This event is emitted when the installation is prepared. This event contains the `installationId` and the handler will create a new `PluginEntity` entity with the `installationId` as the id.
- `handleNumberStored`: Handles the `NumberStored` event. And stores the number in the `pluginEntity` entity in the `number` property.

## Testing

This template uses [matchstick-as](https://github.com/LimeChain/matchstick) framework for unit testing.

> Similar to the `src` folder, the `test` folder contains a `plugin` and `osx` subfolder where the tests for the event handlers from the previous section are written.

Reviewer: ```suggestion
Similar to the `src` folder, the `test` folder contains a `plugin` and `osx` subfolder, where the tests for the event handlers from the previous section are written.
```

- `99_verification/10_verify-contracts.ts`
  - verifies all deployed contracts

# Subgraph Template Usage Guide

## Subgraph dependency installation

Before you begin, make sure you installed the necessary dependencies. This template uses `yarn` so make sure you have it installed. You can install it by running the following command:

```bash
npm install -g yarn
```

After that you can install the dependencies by running:

```bash
cd packages/subgraph/
```

```bash
yarn install
```

> 

Reviewer: This part duplicates the README.md. Let's not introduce more duplication. Take a look what already exists and reference existing texts.

The files containing the replacement values are stored in `manifest/data`. Here you can find a JSON file for each network. The name of the file is the name of the network. For example, the file for the mainnet network is `mainnet.json`. This file contains a property called `dataSources` which holds the information about the contract that will be indexed. In this case, we are only indexing one contract but you can index multiple contracts by adding more objects to the `dataSources` object.

### `schema.graphql`

The second step to create a subgraph is to define the schema. The schema is defined in the `schema.graphql` file located in the `subgraph` folder. This file contains the GraphQL schema that will be used for querying the subgraph. The schema is defined using the GraphQL Schema Definition Language (SDL). You can find more information about it [here](https://graphql.org/learn/schema/).

Here you can add your custom entities and how they are related to each other. You can find more information about the schema [here](https://thegraph.com/docs/define-a-subgraph#defining-the-schema).

## Handlers

> The handling functions will be handling the events specified in the `subgraph.yaml` file. The functions are located in the `src` folder. For this specific case we have 2 handlers:

Reviewer: ```suggestion
Events handler functions are associated with the events in the `subgraph.yaml` file and defined in the `src` folder. For the `MyPlugin` example, two handlers are already provided:
```

> - `handleInstallationPrepared`: Handeles the `InstallationPrepared` event. This event is emitted when the installation is prepared. This event contains the `installationId` and the handler will create a new `PluginEntity` entity with the `installationId` as the id.

Reviewer: ```suggestion
- `handleInstallationPrepared`: Handles the `InstallationPrepared` event. This event is emitted when the installation is prepared. This event contains the `installationId` and the handler will create a new `PluginEntity` entity with the `installationId` as the id.
```

Is all the information provided here necessary to use the template? Just listing the event fields doesn't add much value to the Guide, IMO.

Please go throught the text once more with this mindset.
